Localize backups date and add Choose another button

BrayanDSO opened the pull request:
## Purpose / Description
Backup options were backup name files, which isn't human-readable

## Fixes
Fixes #10362

## Approach
- Refactor part of `BackupManager.kt` in order to work properly and be more clear
    - Other parts of it still need refactoring, but probably are best suited in another PR
- Parse the backup filenames as localizable time strings
- Add a `Choose another` button, so the user can import backups which are not on backups folder

## How Has This Been Tested?

### Human-readable options
https://user-images.githubusercontent.com/69634269/155613612-830cabeb-343f-4b88-8950-8aa5db3062a1.mp4

### Choose another button
https://user-images.githubusercontent.com/69634269/155613828-7c6fca48-3db0-4808-ae91-b8af973e73a6.mp4

## Checklist
_Please, go through these checks before submitting the PR._

- [X] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [X] You have a descriptive commit message with a short title (first line, max 50 chars).
- [X] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [X] You have commented your code, particularly in hard-to-understand areas
- [X] You have performed a self-review of your own code
- [X] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

BrayanDSO commented:
If this gets accepted, please add these context images on Crowdin to help translators:

### restore_backup_date_pattern

![pattern](https://user-images.githubusercontent.com/69634269/155614315-9fd1b68d-d0dc-4be6-9a84-0459d8d712e4.png)

![context seconds](https://user-images.githubusercontent.com/69634269/155614306-79d065c0-201c-48de-b3d8-7b4769c01f8f.png)

### restore_backup_choose_another

![Choose another](https://user-images.githubusercontent.com/69634269/155614347-5506ed23-4bab-4cf1-82b0-cb9c8e99373a.png)

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:
```diff
@@ -196,27 +193,32 @@ class DatabaseErrorDialog : AsyncDialogFragment() {
+                    for (backup in mBackups) {
+                        val ts = BackupManager.getBackupTimeStrings(backup.name)
+                        if (ts != null) {
+                            // "00" seconds member is hardcoded until #10388 isn't fixed
+                            // and backup naming scheme isn't updated to be compatible with Anki Desktop
```
tiniest nitpick

```suggestion
                            // "00" seconds member is hardcoded until #10388 is fixed
                            // and backup naming scheme is updated to be compatible with Anki Desktop
```

BrayanDSO commented:
I'll make clear the design decision I've made about adding `seconds` to the options.
- A part of this PR should be localizing Anki desktop names as well, to help people who uses it (e.g. copying their PC folder to their phones), but because of #10388, this couldn't have been done right now
- The most relevant difference about Desktop and Ankidroid is Desktop having `seconds` on its name scheme, which Ankidroid doesn't have
    - Desktop: `backup-year-month-day-hour.minutes.seconds.colpkg`
    - Ankidroid: `collection-year-month-day-hour-minutes-seconds.colpkg`
- If this was possible, it would be necessary to add a `seconds` string to Ankidroid, so the user can distinguish the moment the backup was made correctly
- I'm adding the seconds member now so when #10388, a new string argument isn't added and translators doesn't have the trouble of retranslating the string

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/BackupManager.kt:
```diff
@@ -354,40 +357,50 @@ open class BackupManager {
+
         @JvmStatic
-        fun getBackups(colFile: File): Array<File?> {
+        fun getBackups(colFile: File): Array<File> {
```
since the getLastBackupDate method requires a sorted array to get the actual last backup, and this method returns a sorted array, it would be nice to have a comment on the method indicating that it returns an array of backups sorted by time with newest last (assuming that is it's sort method?)

BrayanDSO commented:
Hmmmmmm. There is a unit test I built which works on my PC, but doesn't on CI. Sounds fun

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/BackupManager.kt:
```diff
@@ -122,37 +110,52 @@ open class BackupManager {
+        val backupFilename: String = try {
+            String.format(Utils.ENGLISH_LOCALE, "collection-%s.colpkg", df.format(cal.time))
+        } catch (e: UnknownFormatConversionException) {
+            Timber.e(e, "performBackup: error on creating backup filename")
```
```suggestion
            Timber.w(e, "performBackup: error on creating backup filename")
```

You seem to be recovering from the error with a return false above, we try to use error for things that will likely require an app restart - something basically unrecoverable. Warning seems about right as a log level here

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/BackupManager.kt:
```diff
@@ -122,37 +110,52 @@ open class BackupManager {
+    }
+
+    /**
+     * Returns date of the file at last position on files parameter
+     * It doesn't intend to return the *latest* date.
+     * In order to get the latest date, the given files array must be sorted
```
I think if the comments on all your methods used `@param` and `@returns` then you would get better editor behavior when future developers hovered over them in editor? https://kotlinlang.org/docs/kotlin-doc.html#return

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/BackupManager.kt:
```diff
@@ -354,40 +357,50 @@ open class BackupManager {
+         * @return List with {dateformat, year, month, day, hours, minutes} or null if doesn't match naming scheme
+         */
+        @JvmStatic
+        fun getBackupTimeStrings(fileName: String): List<String>? {
+            val m = "collection-((\\d{4})-(\\d{2})-(\\d{2})-(\\d{2})-(\\d{2})).colpkg"
+                .toRegex()
+                .matchEntire(fileName)
+            return m?.groupValues?.subList(1, 7)
+        }
+
         @JvmStatic
-        fun getBackups(colFile: File): Array<File?> {
+        fun getBackups(colFile: File): Array<File> {
             var files = getBackupDirectory(colFile.parentFile).listFiles()
             if (files == null) {
-                files = arrayOfNulls(0)
+                files = arrayOf<File>()
             }
-            val deckBackups = ArrayList<File?>(files.size)
+            val backups = mutableListOf<File>()
```
It would be easier to review this if non-functional refactors like changing the name from "deckBackups" to "backups" were in a completely separate commit, and refactors that are unrelated to the new feature like altering nullability were in a different commit.

As it is, in one commit with refactors mixed in the diff is pretty noisy and it's difficult to focus on just the filename-parsing stuff

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:
```diff
@@ -180,13 +180,10 @@ class DatabaseErrorDialog : AsyncDialogFragment() {
-                }
+                mBackups = BackupManager.getBackups(File(path))
+                // Since backups are sorted ascendingly at getBackups(),
+                // revert it to show latest backups at top
```
```suggestion
                // reverse it to show latest backups at top
```

mikehardy commented on AnkiDroid/src/test/java/com/ichi2/anki/BackupManagerTest.java:
```diff
@@ -42,6 +49,55 @@
+    @Test
+    public void getLastBackupDateTest() throws ParseException {
+        BackupManager bm = BackupManager.createInstance();
+        File[] backups = {
+                new File ("collection-2000-12-31-23-04.colpkg"),
+                new File ("collection-2010-01-02-03-04.colpkg"),
+                new File ("collection-1999-12-31-23-59.colpkg"),
+        };
```
Could you add some invalid names in here to make sure you check handling non-parseable names? And/or in a separate test?

mikehardy reviewed (changes requested):
Seems pretty good - I got pretty detailed on the comments but I think the doc changes will help, and testing failure cases should definitely add confidence in the change

mikehardy commented:
Unit test missed comparison on the seconds it appears, they may not be clamped to 00 as indicated in comments

```
com.ichi2.anki.BackupManagerTest > getNewBackupNameTest FAILED
    org.junit.ComparisonFailure: Backup name doesn't match naming scheme expected:<...lection-1970-01-02-0[0]-46.colpkg> but was:<...lection-1970-01-02-0[3]-46.colpkg>
        at org.junit.Assert.assertEquals(Assert.java:117)
        at com.ichi2.anki.BackupManagerTest.getNewBackupNameTest(BackupManagerTest.java:57)
```

david-allison reviewed (commented):
Echoing Mike, it'd be great to split out the refactoring and functional changes to make this easy to review. I haven't reviewed too hard as the commits weren't split out, but conceptually, I like this.

Would you consider also using [getRelativeTimeSpanString](https://developer.android.com/reference/android/text/format/DateUtils.html#getRelativeTimeSpanString%28long%29) when displaying times (should display: `X mins ago` in an internationalized format)

Anki Desktop now includes the seconds, as well as `backup-` in the backup filename, this would be useful as either a follow-on PR, or an open issue: https://github.com/ankitects/anki/blob/a0d0f2f8fd7cca6f3426c9f28ad8ffdfdbcc648c/qt/aqt/main.py#L653-L689

"getNewBackupName" is a concerning name, as it immediately makes me want to investigate /why/ it's new. I'd just use `getBackupName` with a historical note that before 2022, it produced `XXX`.

I really like the tests, thanks!

BrayanDSO commented:
With the last force push:
- Fixed the issues Mike addressed
- Removed the unnecessary refactors (sorry, I went overboard with it)
- Fixed the CI Unittest problem (I hope)
- Improved the functionality of `getLastBackupDate()`

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/BackupManager.kt:
```diff
@@ -122,37 +112,51 @@ open class BackupManager {
+     * It doesn't intend to return the *latest* date.
+     * In order to get the latest date, the given files array must be sorted
+     */
+    fun getLastBackupDate(files: Array<File>): Date? {
```
This was probably wrong before, but it still doesn't seem correct.

If this is returning an unintuitive result, I'd much rather that this was easier to read. Something along the lines of:


```kotlin
return files
    .reversed()
    .mapNotNull { getBackupTimeStrings(it.name) }
    .mapNotNull { tryParseDate(it[0]) }
    .firstOrNull()

```

BrayanDSO replied:
I agree this is an unintuitive function. Something parsing all file dates and returning the `maxOrNull()` would be better, but unecessary since files are already sorted because `deleteBackups()` needs them sorted anyway

Thanks for the code suggestion, now it looks a lot cleaner

david-allison replied:
I'd much prefer the `maxOrNull` approach if that works for you

BrayanDSO replied:
Initially, I haven't done this because with only one name scheme, sorting by filename is easier and costs less performance than parsing all of them with dateformat then sorting. So I just kept using the same behavior as before, only refactoring what was necessary to build the feature.

In order to get it more clear, I'll refactor `getBackups` and `deleteDeckBackups` as well. `maxOrNull` probably is still unnecessary, but this deserves something more intuitive

BrayanDSO replied:
I think 4bfdaf0 does the job. I'm satisfied with the result.

Please review it when you get the time

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/BackupManager.kt:
```diff
@@ -122,37 +112,51 @@ open class BackupManager {
+        }
+        for (file in files.reversed()) {
+            val ts = getBackupTimeStrings(file.name)
+            if (ts != null) {
```
ts.length also needs to be greater than 0

BrayanDSO replied:
getBackupTimeStrings either returns a `List` with 6 strings or null. If isn't null, is it really necessary to check its length?

david-allison replied:
Sounds fine to me

BrayanDSO commented:
> Would you consider also using [getRelativeTimeSpanString](https://developer.android.com/reference/android/text/format/DateUtils.html#getRelativeTimeSpanString%28long%29) when displaying times (should display: `X mins ago` in an internationalized format)

Do you mean, something like `02-12-2022 18:24 (15 minutes ago)`?
 
> Anki Desktop now includes the seconds, as well as `backup-` in the backup filename, this would be useful as either a follow-on PR, or an open issue: https://github.com/ankitects/anki/blob/a0d0f2f8fd7cca6f3426c9f28ad8ffdfdbcc648c/qt/aqt/main.py#L653-L689

When I tried to get both name schemes working, I updated the file name to equal to anki desktop scheme too. 
I think it's better to not update it yet, because people may copy their anki folders from desktop and try to restore a backup, but they won't be able to because of #10388. So, to avoid this problem, it's better to keep using the current filename pattern

> "getNewBackupName" is a concerning name, as it immediately makes me want to investigate /why/ it's new. I'd just use `getBackupName` with a historical note that before 2022, it produced `XXX`.

I changed the name because when I first read it, I didn't understand what it meant to do. It gets the name of an existing backup or it creates a new backup name? So I changed to `getNewBackupName`, you call it to get a new backup name.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/BackupManager.kt:
```diff
@@ -354,40 +358,54 @@ open class BackupManager {
+         */
+        @JvmStatic
+        fun getBackupTimeStrings(fileName: String): List<String>? {
+            val m = "collection-((\\d{4})-(\\d{2})-(\\d{2})-(\\d{2})-(\\d{2})).colpkg"
```
Move this to a static, rather than recreating a regex, for performance.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/BackupManager.kt:
```diff
@@ -122,37 +112,51 @@ open class BackupManager {
+     * It doesn't intend to return the *latest* date.
+     * In order to get the latest date, the given files array must be sorted
+     */
+    fun getLastBackupDate(files: Array<File>): Date? {
```
This was probably wrong before, but it still doesn't seem correct.

If this is returning an unintuitive result, I'd much rather that this was easier to read. Something along the lines


````
return files
    .reversed()
    .mapNotNull { getBackupTimeStrings(it.name) }
    .mapNotNull { tryParseDate(it[0]) }
    .firstOrNull()

```

david-allison reviewed (commented):
This could still do with a few `edit rebases` to split out commits. This is a partial review. 

Please pushback on the comments here, you've read this more than I have.

david-allison commented:
> Do you mean, something like `02-12-2022 18:24 (15 minutes ago)`?

Looks great!

> When I tried to get both name schemes working, I updated the file name to equal to anki desktop scheme too. I think it's better to not update it yet, because people may copy their anki folders from desktop and try to restore a backup, but they won't be able to because of #10388. So, to avoid this problem, it's better to keep using the current filename pattern

Understood!

> I changed the name because when I first read it, I didn't understand what it meant to do. It gets the name of an existing backup or it creates a new backup name? So I changed to `getNewBackupName`, you call it to get a new backup name.

Hmm... `getNameOfNewBackup`

BrayanDSO commented:
```
Would you consider also using [getRelativeTimeSpanString](https://developer.android.com/reference/android/text/format/DateUtils.html#getRelativeTimeSpanString%28long%29) when displaying times (should display: X mins ago in an internationalized format)
```
If the horizontal screen space was larger, this would be great with a right aligned smaller grey text, like this mockup:

![2022_02_24-21_25_48](https://user-images.githubusercontent.com/69634269/155630446-edd378cd-fe40-4a3f-a734-854062d5cc6d.png)

But since people use their phones mostly on portrait mode, it would be polluted to have bigger strings.

With some vertical spacing, it would look fine too, like this other mockup:

![2022_02_24-21_32_27](https://user-images.githubusercontent.com/69634269/155630849-37404f0c-7c03-4a8c-aefc-35c9a77cdb90.png)

But since most of the work would be in the dialog styling, I think it's better to adress it in another issue and PR.
Now that I did this last mockup, it looks better than I expected. Maybe I'll work on it later.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/BackupManager.kt:
```diff
@@ -139,15 +138,24 @@ open class BackupManager {
     fun getLastBackupDate(files: Array<File>): Date? {
         return files
-            .reversed()
+            .sortedDescending()
```
Don't you want to sort after parsing so we select the last "known" backup

BrayanDSO replied:
It makes sense. The most clean solution would be:
```kotlin
return files.
    .mapOrNull { getBackupTimeStrings(it.name) }
    .mapOrNull { getBackupDate(it[0])}
    .maxOrNull()
```
Thinking more about it, I'm inclined to do something different to avoid unnecessary RegEx and dateparsing processing 
(pseudo code)
```
for (i in files.sortedDescending) {
    return try {
         df.parse(file.name)
    } catch {}
}
return null
```
When I get some time with my PC, I'll try a cleaner solution based on the pseudocode

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/BackupManager.kt:
```diff
@@ -354,40 +359,50 @@ open class BackupManager {
         @JvmStatic
-        fun getBackups(colFile: File): Array<File?> {
+        fun getBackupTimeStrings(fileName: String): List<String>? {
+            val m = backupNameRegex.matchEntire(fileName)
```
If I'm understanding correctly: This no longer lists backups under the "old" naming styles for deletion, requiring significantly more space.

A test here would be useful

BrayanDSO replied:
I don't get what you mean. This is what currently happens:
- `deleteDeckBackups()` get their  files by calling `getBackups()`.
- `getBackups()` uses `getBackupTimeStrings` just to check if result isn't null, what means the filename is a valid backup and should be added to backups list

BrayanDSO replied:
I'll use `matchEntire` instead of `getBackupTimeStrings` in `getBackups` to improve performance, if it's that what you meant

BrayanDSO replied:
I'll build a test to guarantee `getBackups()` functionality when I get the time

BrayanDSO replied:
Done. Waiting for feedback

david-allison reviewed (changes requested):
I think this breaks backup deletion

BrayanDSO commented:
> I think this breaks backup deletion

It doesn't. But I'll make some tests for safety

BrayanDSO commented:
`getBackups()` and `deleteDeckBackups()` now have tests too and `getLastBackupDate` is working as intended.

A question: when it is necessary to create files on tests, is `deleteOnExit()` necessary as well? 
I assumed and checked on my PC the tests were realized on a temp folder then deleted even without the command, but I can't be sure how this works on another systems

david-allison commented:
> `getBackups()` and `deleteDeckBackups()` now have tests too and `getLastBackupDate` is working as intended.
> 
> A question: when it is necessary to create files on tests, is `deleteOnExit()` necessary as well? I assumed and checked on my PC the tests were realized on a temp folder then deleted even without the command, but I can't be sure how this works on another systems

`createTransientFile` should work. I didn't realise earlier, but TMP isn't guaranteed to be cleared on Linux.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/BackupManager.kt:
```diff
@@ -216,14 +219,16 @@ open class BackupManager {
         private const val MIN_BACKUP_COL_SIZE = 10000 // threshold in bytes to backup a col file
         private const val BACKUP_SUFFIX = "backup"
         const val BROKEN_DECKS_SUFFIX = "broken"
+        private val backupNameRegex = Regex("collection-((\\d{4})-(\\d{2})-(\\d{2})-(\\d{2})-(\\d{2})).colpkg")
```
The previous code looked for `.apkg` as well, is there a reason this was removed?

BrayanDSO replied:
1. Anki desktop only accepts `.colpkg` as backup
2. Replacing the entire collection with `.apkg` isn't a expected behavior for the user

david-allison commented on AnkiDroid/src/main/res/values/03-dialogs.xml:
```diff
@@ -106,6 +106,8 @@
 
     <string name="restore_backup_title">Restore backup?</string>
     <string name="restore_backup">Your collection will be replaced with an older copy. Any unsaved progress will be lost.</string>
+    <string name="restore_backup_date_pattern">%2$s-%3$s-%1$s, %4$s:%5$s:%6$s</string>
```
This is MM-DD-YYYY which is more confusing for European users.

david-allison reviewed (changes requested):
This is less localized in terms of EU/US date formats.

I /think/ `DateFormat.getBestDateTimePattern()` would work here: we want the setting from the user's phone, not one from our translators

BrayanDSO commented:
> https://developer.android.com/reference/android/text/format/DateFormat#getBestDateTimePattern(java.util.Locale,%20java.lang.String)

![](https://media.giphy.com/media/p8GJOXwSNzQPu/giphy.gif)

### Some results

![Screenshot_20220227-084919_AnkiDroid](https://user-images.githubusercontent.com/69634269/155881394-61858af8-d81c-4a37-acb7-5f962d0ae301.png)
![Screenshot_20220227-084946_AnkiDroid](https://user-images.githubusercontent.com/69634269/155881395-6a28b155-d573-4f70-9c5e-79aedfe51c53.png)
![Screenshot_20220227-085528_AnkiDroid](https://user-images.githubusercontent.com/69634269/155881396-125470e0-cd46-4cc2-94bd-2b767990d777.png)


With Ankidroid `English` translation, it looks that `getBestDateTimePattern` defaults to American as well. 

It works properly when using `System default` with my system on UK english.

Therefore, I believe the `English` translation option should be renamed to `English (United States)` or something like it.

BrayanDSO commented:
> `createTransientFile` should work. I didn't realise earlier, but TMP isn't guaranteed to be cleared on Linux.

The test needs file with parseable names. Since `createTransientFile` doesn't accept filenames, this doen't work.

I think making a `createNamedTransientFile` would be good.

The major "problem" would be deleting the filetree `getBackupDirectory` requires, which will probably make some ugly code, but I'll see what I can do

david-allison reviewed (commented):
For discussion with other maintainers.

I feel we should be using the User's system datetime rather than the AnkiDroid language here. 

That would mean saving the original locale before `updateContextWithLanguage` is called in `AnkiDroidApp`.

I'd like to see about what others think here, could you drop a line in the Discord to see what others think. I don't want to ask for work which will be reverted, and it's a complex piece to get right.

BrayanDSO commented:
Making it more clear:

![Screenshot_20220227-131259_AnkiDroid](https://user-images.githubusercontent.com/69634269/155890247-9c97536b-2cbb-4031-be8c-58c6d68115b5.png)

When using `English` (selected option), the localization defaults to american dates.

When using `System language` (first option) and the system is on UK english, the localization works properly with UK dates.
(same logic may be appliable to other countries language variations, like spanish and french)

BrayanDSO commented:
**My opinion** is:

I don't use Ankidroid with `System language`, because I prefer `english` (earlier I even used japanese too) and my system language is portuguese. 

If I saw a portuguese (my system language) date scheme (`dd-mm-yyyy`) while using english language (date scheme `mm-dd-yyy`), I'd be very confused, since they may clash with each other (is `10-02-2022`  `February 10, 2022` or `October 2, 2022`?).

If I explicitly chose not to use my system language on Ankidroid, I wouldn't expect it to do it in just a specific place of the app.

Therefore, I'd rather renaming `English` option to `English (United States)` (and maybe some other countries who has this ambiguity), because even without this PR it shouldn't be assumed that all `english` are the same

david-allison commented:
> **My opinion** is:
> 
> I don't use Ankidroid with `System language`, because I prefer `english` (earlier I even used japanese too) and my system language is portuguese.
> 
> If I saw a portuguese date scheme (`dd-mm-yyyy`) while using english language (date scheme `mm-dd-yyy`), I'd be very confused, since they may clash with each other (is `10-02-2022` `February 10, 2022` or `October 2, 2022`?).
> 
> If I explicitly chose not to use my system language on Ankidroid, I wouldn't expect it to do it in just a specific place of the app.
> 
> Therefore, I'd rather renaming `English` option to `English (United States)` (and maybe some other countries who has this ambiguity), because even without this PR it shouldn't be assumed that all `english` are the same

Is there a reason we're not using the named month `MMM` in the `skeleton` parameter?

BrayanDSO commented:
Yes. It may more volatile since there are many translations. I can't say if there won't be strings which take more than one line and that the behavior will be good in all of the translations.

It would be much more readable with `MMM` IMO, so I'll test with ~10 languages to check it

BrayanDSO commented:
Definetly better (skeleton `dd MMM yyyy HH:mm`)

![Screenshot_20220227-141149_AnkiDroid](https://user-images.githubusercontent.com/69634269/155892290-e124f67a-bf55-4294-8b2e-b4f4517f896d.png)
![Screenshot_20220227-140813_AnkiDroid](https://user-images.githubusercontent.com/69634269/155892291-3efe6242-2051-4d7b-9df7-9f09d3fc11e1.png)
![Screenshot_20220227-140855_AnkiDroid](https://user-images.githubusercontent.com/69634269/155892287-6e6f88e3-b97c-441e-905d-892f45cea8e6.png)
![Screenshot_20220227-140924_AnkiDroid](https://user-images.githubusercontent.com/69634269/155892288-8202400c-ecd3-4dfa-a361-0bf5baba80f3.png)
![Screenshot_20220227-141117_AnkiDroid](https://user-images.githubusercontent.com/69634269/155892289-5b7e3aa0-1bb5-4bd5-aebd-603e0aac472c.png)

viciousAegis commented:
>Therefore, I'd rather renaming English option to English (United States) (and maybe some other countries who has this ambiguity), because even without this PR it shouldn't be assumed that all english are the same

I agree here, there are a lot of different versions of English used depending on the country, which have subtle changes. It will be good to rename and add the other versions of English there. 

Regarding the skeleton, `dd MMM yyyy HH:mm` definitely removes a lot of ambiguity and is the much better choice. Nice work on this! I know a few languages on there and it makes it crystal clear in each, I like this version a lot.

david-allison reviewed (approved):
This looks great! 

It'll need to be `squash`ed when it goes in, which makes it a large commit.

~~My one remaining question: has `choose another` been tested?~~

BrayanDSO commented:
> My one remaining question: has `choose another` been tested?

The demo video is on the PR text at `How Has This Been Tested?`

mikehardy commented:
Just chiming in to echo this:

> If I saw a portuguese date scheme (dd-mm-yyyy) while using english language (date scheme mm-dd-yyy), I'd be very confused, since they may clash with each other (is 10-02-2022 February 10, 2022 or October 2, 2022?).

I use Spanish or English locales for apps and devices depending on what my goal is and my brain is expecting the Spanish date formats (and number formats) when I'm in the Spanish language. it would be really weird not to translate the date / time / number formats as well, in my opinion

localization is the whole group of LTR, actual text symbols in use, numeric formatting, pluralization - all of it, and it should be consistent I think

mikehardy commented on AnkiDroid/src/test/java/com/ichi2/anki/BackupManagerTest.java:
```diff
@@ -42,6 +52,119 @@
+        File[] backups = BackupManager.getBackups(colFile);
+
+        assertNotNull(backups);
+        assertArrayEquals("Only the valid backup names should have been kept", new File[] {f1, f3}, backups);
```
Here's the location of test failure on my local machine. It appears f3 is first, not f1, not sure why. Either sorting, or f1 and f2 were removed instead of just f2? Surprised this passes in CI, our unit tests are running on a linux runner as well so it fairly closely matches my local setup where I see the failure every run :thinking:

mikehardy reviewed (changes requested):
unit test failure for me as noted, otherwise code reviews great - definitely need to identify the root cause of the unit test failure prior to merge tho

BrayanDSO commented:
Try these alternative versions please. Since it works on my PC and Github CI, I don't know how else I can help

First:
```java
    @Test
    @SuppressWarnings("ResultOfMethodCallIgnored")
    public void getBackupsTest() throws IOException {
        // getBackups() doesn't require a proper collection file
        // because it is only used to get its parent
        File colFile = tempFolder.newFile();
        assertEquals(0, BackupManager.getBackups(colFile).length);

        File backupDir = BackupManager.getBackupDirectory(tempFolder.getRoot());
        File f1 = new File (backupDir, "collection-2000-12-31-23-04.colpkg");
        File f2 = new File (backupDir, "foo");
        File f3 = new File (backupDir, "collection-2010-12-06-13-04.colpkg");
        f1.createNewFile();
        f2.createNewFile();
        f3.createNewFile();
        File[] backups = BackupManager.getBackups(colFile);

        assertNotNull(backups);
        assertEquals("Only the valid backup names should have been kept", 2, backups.length);
        assertEquals("collection-2000-12-31-23-04.colpkg", backups[0].getName());
        assertEquals("collection-2010-12-06-13-04.colpkg", backups[1].getName());
    }
```
Second:
```java
import java.util.Arrays;
import java.util.List;
```
```java
    @Test
    @SuppressWarnings("ResultOfMethodCallIgnored")
    public void getBackupsTest() throws IOException {
        // getBackups() doesn't require a proper collection file
        // because it is only used to get its parent
        File colFile = tempFolder.newFile();
        assertEquals(0, BackupManager.getBackups(colFile).length);

        File backupDir = BackupManager.getBackupDirectory(tempFolder.getRoot());
        File f1 = new File (backupDir, "collection-2000-12-31-23-04.colpkg");
        File f2 = new File (backupDir, "foo");
        File f3 = new File (backupDir, "collection-2010-12-06-13-04.colpkg");
        f1.createNewFile();
        f2.createNewFile();
        f3.createNewFile();
        File[] backups = BackupManager.getBackups(colFile);

        assertNotNull(backups);
        assertEquals("Only the valid backup names should have been kept", 2, backups.length);
        List<File> backupsList = Arrays.asList(backups);
        assertTrue(backupsList.contains(f1));
        assertFalse(backupsList.contains(f2));
        assertTrue(backupsList.contains(f3));
    }
```

mikehardy commented:
I still haven't had time to actually investigate but I saw you pushed changes so I just pulled them and reran with same result, won't have time to investigate for a while unfortunately

```
com.ichi2.anki.BackupManagerTest > getBackupsTest FAILED
    Only the valid backup names should have been kept: arrays first differed at element [0]; expected:</tmp/junit17950458880720144227/backup/collection-2000-12-31-23-04.colpkg> but was:</tmp/junit17950458880720144227/backup/collection-2010-12-06-13-04.colpkg>
        at org.junit.internal.ComparisonCriteria.arrayEquals(ComparisonCriteria.java:78)
        at org.junit.internal.ComparisonCriteria.arrayEquals(ComparisonCriteria.java:28)
        at org.junit.Assert.internalArrayEquals(Assert.java:534)
        at org.junit.Assert.assertArrayEquals(Assert.java:285)
        at com.ichi2.anki.BackupManagerTest.getBackupsTest(BackupManagerTest.java:143)

        Caused by:
        java.lang.AssertionError: expected:</tmp/junit17950458880720144227/backup/collection-2000-12-31-23-04.colpkg> but was:</tmp/junit17950458880720144227/backup/collection-2010-12-06-13-04.colpkg>
            at org.junit.Assert.fail(Assert.java:89)
            at org.junit.Assert.failNotEquals(Assert.java:835)
            at org.junit.Assert.assertEquals(Assert.java:120)
            at org.junit.Assert.assertEquals(Assert.java:146)
            at org.junit.internal.ExactComparisonCriteria.assertElementsEqual(ExactComparisonCriteria.java:8)
            at org.junit.internal.ComparisonCriteria.arrayEquals(ComparisonCriteria.java:76)
            ... 4 more

```

BrayanDSO commented:
> I still haven't had time to actually investigate but I saw you pushed changes so I just pulled them and reran with same result, won't have time to investigate for a while unfortunately
> 
> ```
> com.ichi2.anki.BackupManagerTest > getBackupsTest FAILED
>     Only the valid backup names should have been kept: arrays first differed at element [0]; expected:</tmp/junit17950458880720144227/backup/collection-2000-12-31-23-04.colpkg> but was:</tmp/junit17950458880720144227/backup/collection-2010-12-06-13-04.colpkg>
>         at org.junit.internal.ComparisonCriteria.arrayEquals(ComparisonCriteria.java:78)
>         at org.junit.internal.ComparisonCriteria.arrayEquals(ComparisonCriteria.java:28)
>         at org.junit.Assert.internalArrayEquals(Assert.java:534)
>         at org.junit.Assert.assertArrayEquals(Assert.java:285)
>         at com.ichi2.anki.BackupManagerTest.getBackupsTest(BackupManagerTest.java:143)
> 
>         Caused by:
>         java.lang.AssertionError: expected:</tmp/junit17950458880720144227/backup/collection-2000-12-31-23-04.colpkg> but was:</tmp/junit17950458880720144227/backup/collection-2010-12-06-13-04.colpkg>
>             at org.junit.Assert.fail(Assert.java:89)
>             at org.junit.Assert.failNotEquals(Assert.java:835)
>             at org.junit.Assert.assertEquals(Assert.java:120)
>             at org.junit.Assert.assertEquals(Assert.java:146)
>             at org.junit.internal.ExactComparisonCriteria.assertElementsEqual(ExactComparisonCriteria.java:8)
>             at org.junit.internal.ComparisonCriteria.arrayEquals(ComparisonCriteria.java:76)
>             ... 4 more
> ```

This is the first test. Does any of `getBackupsTest2`, `getBackupsTest3` or `getBackupsTest4` work?

david-allison commented:
Issue:

`listFiles()` order isn't guaranteed and is called in `getBackups()`

`sort(deckBackups)` is no longer called in `getBackups()`.

Either we don't care, in which case we want to perform an unordered comparison in the test, or we've introduced a bug and broken an assumption.

mikehardy commented:
@BrayanDSO it does not appear that you have addressed the part about "have we broken an assumption here" that David mentioned in his comment https://github.com/ankidroid/Anki-Android/pull/10396#issuecomment-1054890556

When I look at the files now on github web UI it appears you have altered only the test.

Please carefully examine this statement and respond with your thoughts:

> [](https://github.com/BrayanDSO)Either we don't care, in which case we want to perform an unordered comparison in the test, or we've introduced a bug and broken an assumption.

If we do not care about order of backups in presentation then I want to know why - I was under the impression we do want to present an ordered list? Or is this really just a test-only issue that on my machine they appear to come out unordered and break the test?

BrayanDSO commented:
> Or is this really just a test-only issue that on my machine they appear to come out unordered and break the test?

↑ `getBackups()` don't need to be sorted in its implementation. On my PC and Github CI it follows the default sorting order. But since on your computer and maybe anothers it may not, I added a `sort()` on the test so the assertions can be done correctly. Probably a comment there would be good to explain the `sort()` reason

mikehardy reviewed (approved):
This works for me now in local testing
